go/libraries/doltcore/sqle/expeval/expression_evaluator.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expeval
+
+import (
+	"context"
+
+	"github.com/src-d/go-mysql-server/sql"
+	"github.com/src-d/go-mysql-server/sql/expression"
+	"gopkg.in/src-d/go-errors.v1"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+var errUnsupportedComparisonType = errors.NewKind("Unsupported Comparison Type.")
+var errUnknownColumn = errors.NewKind("Column %s not found.")
+var errInvalidConversion = errors.NewKind("Could not convert %s from %s to %s.")
+var errNotImplemented = errors.NewKind("Not Implemented: %s")
+
+// ExpressionFunc is a function that takes a map of tag to value and returns whether some set of criteria are true for
+// the set of values
+type ExpressionFunc func(ctx context.Context, vals map[uint64]types.Value) (bool, error)
+
+// ExpressionFuncFromSQLExpressions returns an ExpressionFunc which represents the slice of sql.Expressions passed in
+func ExpressionFuncFromSQLExpressions(nbf *types.NomsBinFormat, sch schema.Schema, expressions []sql.Expression) (ExpressionFunc, error) {
+	var root ExpressionFunc
+	for _, exp := range expressions {
+		expFunc, err := getExpFunc(nbf, sch, exp)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if root == nil {
+			root = expFunc
+		} else {
+			root = newAndFunc(root, expFunc)
+		}
+	}
+
+	return root, nil
+}
+
+func getExpFunc(nbf *types.NomsBinFormat, sch schema.Schema, exp sql.Expression) (ExpressionFunc, error) {
```

Comment:
So this is a pretty good first pass at implementing a dolt-native expression tree evaluator. But it significantly duplicates the machinery that does basically the exact same thing, but for abstract rows, in go-mysql-server. And it is less complete than the rest of the query engine, e.g. col < 2+2 isn't handled.

Here's a crazy thought: what if instead of doing it this way, you implemented a few classes that implement sql.Expression and then replace the GetField nodes in the expression tree with your new wrapper type? Then you just call Eval() on the root of the expression tree, and all the usual go-mysql-server evaluation logic just works. Because the sql.Row type is just []interface{}, you could just pass the row's tag map to Eval(). Then all you need is some code to extract the noms value at a particular tag and convert it to a particular primitive value type and the evaluation engine does the rest, including all the hairy type conversions.

If you do that, then you can call expressions.TransformUp() on this expression tree, replacing GetField expressions with your custom field grabber and wrapping every other kind of expression with something that transforms the map[tag]field into the slice of primitive values the rest of the evaluation engine expects.

I can imagine this working, but there are probably some hairy details. But there are big advantages in consistency and completeness to doing it this way. Worth a shot, I think.